karoo_gp/base_class.py

```diff
@@ -14,182 +14,162 @@
+    │   └─ .evolve()                    - return a new generation of trees
+    │
+    ├─ .fit(X, y)                       - evolve expressions to predict y
+    ├- .predict(X)                      - return predicted y values for X
+    └- .score(pred, y)                  - return score of prediction against y
+    """
+
+    def __init__(
+        self, tree_type='r', tree_depth_base=3, tree_depth_max=3,
+        tree_depth_min=1, tree_pop_max=100, gen_max=10, tourn_size=7,
+        filename='', output_dir='', evolve_repro=0.1, evolve_point=0.1,
+        evolve_branch=0.2, evolve_cross=0.6, display='s', precision=None,
+        swim='p', mode='s', seed=None, pause_callback=None,
+        engine_type='numpy', tf_device="/gpu:0", tf_device_log=False,
+        functions=None, terminals=None, test_size=0.2, scoring=None,
+        higher_is_better=False, _cache={}):
+        """Initialize a Karoo_GP object with given parameters"""
+
+        # Model parameters
         self.seed = seed
-        self.rng = np.random.default_rng(seed)  # this is used by Karoo
-        # the following two are not used, set them just in case
-        np.random.seed(seed)  # this was used by sklearn while classifying
-        tf.set_random_seed(seed)
-
-        ### PART 2 - construct first generation of Trees ###
-        self.fx_data_load(filename)
-        self.functions = Functions([f[0] for f in self.functions])  # Symbol only
-        self.terminals = Terminals(self.terminals[:-1])
-
-        self.fx_fitness_labels_map = fx_fitness_labels_map_maker(self.class_labels)
-
-        self.log(f'\n\t\033[32m Press \033[36m\033[1m?\033[0;0m\033[32m at any '
-                 f'\033[36m\033[1m(pause)\033[0;0m\033[32m, or '
-                 f'\033[36m\033[1mENTER\033[0;0m \033[32mto continue the run\033[0;0m',
-                 display=['i'])
-        self.pause(display=['i'])
-
-        # initialise population_a to host the first generation
+        self.rng = np.random.default_rng(seed)
+        np.random.seed(seed)                 # used by skm in train_test_split
+        self.cache = _cache                   # scores by hash(data), expression
+        self.datetime = datetime.now().strftime('%Y-%m-%d_%H-%M-%S-%f')
+        self.filename = filename             # prefix for output files
+        self.output_dir = output_dir         # path to output directory
+        self.loader = DataLoader(self)       # set path, initialize logs
+        self.mode = mode                     # determines if pauses after fit
+        self.display = display               # determines when pause/log called
+        self.pause_callback = pause_callback # called throughout based on disp
+        self.history = {}                    # best score for each field/gen
+
+        # Initialize Population
+        self.functions = Functions(functions) if functions is not None else Functions.arithmetic()
+        self.terminals = terminals if type(terminals) is Terminals else Terminals(terminals)
```

Comment:
This is another case of an input that accepts different type that then needs to be "normalized".  Is there a reason to support them?
```suggestion
        self.terminals = terminals if isinstance(terminals, Terminals) else Terminals(terminals)
```

Reply:
Here and `Functions`, ya I think so.